### PR TITLE
Laurie/msg gen browser

### PR DESCRIPTION
**Pull request by LaurieCheers-unity**
Created a nicer UI for building .msg and .srv files.
Just open RosMessageGeneration->Browse, tell it where your ROS src folder is, and then you can navigate around and click to build individual files or entire folders.

### CODE REVIEW

File: Editor/MessageGeneration/MessageGenBrowser.cs [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+            window.maxSize = new Vector2(800, 1200);
+            window.Show();
+        }
+
+        protected virtual void OnGUI()
+        {
+            EditorGUILayout.BeginHorizontal();
+            MessageGenBrowserSettings settings = MessageGenBrowserSettings.Get();
+            string inPath = settings.inputPath;
+            inPath = EditorGUILayout.TextField("Input Path", inPath);
```
Comment (at669) [inline Editor/MessageGeneration/MessageGenBrowser.cs:58]: can this title be clarified so that this should be the top-level ROS workspace folder?

**Review by at669 — COMMENTED** [review]
this is great! a lot easier than having to navigate to multiple directories. 

- srv button says `msgs`
- button to generate directory of services doesn't work for me, but each individual one does
Screen Shot 2020-12-08 at 10 45 23 AM

File: Editor/MessageGeneration/MessageGenBrowser.cs (hunk 1 above)
Comment (LaurieCheers-unity) [reply Editor/MessageGeneration/MessageGenBrowser.cs:58]: > this is great! a lot easier than having to navigate to multiple directories.

Thanks! Yeah I've had to regenerate the common_msgs folder a couple of times, this is definitely my attempt to make that less annoying :-)

The title could be clearer but I'm not sure what exactly to put - it doesn't have to be the top level workspace folder, just literally any folder that contains messages somewhere.

**Comment by mpinol** [comment]
`MessageTokenizerException: Unexpected token '--- '. Did you mean '---' (ROS Service/Action seperator)?` Just ran into this error when trying to generate a service with a trailing space after the line break. Tested on `dev` and it looked like everything generated just fine.

File: Editor/MessageGeneration/MessageGenBrowser.cs [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+                case CachedEntryStatus.UnbuiltMsgFile:
+                    buildLabel = "Build msg";
+                    break;
+                case CachedEntryStatus.UnbuiltSrvFile:
+                    buildLabel = "Build srv";
+                    break;
+                case CachedEntryStatus.NormalFolder:
+                case CachedEntryStatus.MsgFolder:
+                    isFolder = true;
+                    buildLabel = (entry.numMsgs != null) ? "Build " + entry.numMsgs : null;
```
Comment (mrpropellers) [inline Editor/MessageGeneration/MessageGenBrowser.cs:127]: What does it mean for buildLabel to be null here?  Is that an error case?  Should it return?

File: Editor/MessageGeneration/MessageGenBrowser.cs [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+
+        [SerializeField]
+        private List<string> foldedOutList = new List<string> { };
+        [SerializeField]
+        private Vector2 scrollPos;
+
+        const int BUTTON_WIDTH = 100;
+
+        bool isCacheDirty = true;
+        CachedEntry cacheRoot;
```
Comment (mrpropellers) [inline Editor/MessageGeneration/MessageGenBrowser.cs:22]: Some private variables are getting `private` labels and others aren't. They could also use a prefix to differentiate them from local variables. Standard Unity style would be to exclude `private` and use `m_CapsCase`

File: Runtime/MessageGeneration/MessageAutoGen.cs [hunk 4]
```diff
@@ -9,16 +9,34 @@ namespace RosMessageGeneration
 {
     public class MessageAutoGen
     {
+        public static string GetRosPackageName(string inPath, string rosPackageNameOverride = "")
+        {
+            if (!rosPackageNameOverride.Equals(""))
+                return rosPackageNameOverride;
+
+            string[] hierarchy = inPath.Split(new char[] { '/', '\\' });
+            return hierarchy[hierarchy.Length - 3];
```
Comment (mrpropellers) [inline Runtime/MessageGeneration/MessageAutoGen.cs:18]: This seems to be making an assumption about a very specific file hierarchy inside the catkin_package. Message files can be pretty much anywhere, though, as long as they are declared correctly in the CMakeLists. The only guaranteed way to determine the package name is to find the package.xml or CMakeLists.txt at the package root and read it from there.

**Review by mrpropellers — CHANGES_REQUESTED** [review]

File: Editor/MessageGeneration/MessageGenBrowser.cs (hunk 2 above)
Comment (LaurieCheers-unity) [reply Editor/MessageGeneration/MessageGenBrowser.cs:127]: That means this line doesn't have a "Build" button.

File: Editor/MessageGeneration/MessageGenBrowser.cs [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+                    buildLabel = (entry.numMsgs != null) ? "Build " + entry.numMsgs : null;
+                    break;
+            }
+
+            if(!isFolder)
+            {
+                EditorGUILayout.BeginHorizontal();
+                EditorGUILayout.LabelField(entry.label);
+
+                if(buildLabel != null)
```
[the remark this answers is not in the document]
Comment (LaurieCheers-unity) [reply Editor/MessageGeneration/MessageGenBrowser.cs:136]: Could do, I guess, but I wasn't envisioning it as a critical failure - I'd interpret it as meaning the cache system decided this should display as a file with no build button. At the moment we never do that, but we could.

File: Runtime/MessageGeneration/MessageAutoGen.cs (hunk 4 above)
Comment (LaurieCheers-unity) [reply Runtime/MessageGeneration/MessageAutoGen.cs:18]: Hmm, good to know. This is the logic we were already using to identify the package name... I guess it needs some revision.

File: Editor/MessageGeneration/MessageGenBrowser.cs [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+using System.Linq;
+using UnityEditor;
+using UnityEngine;
+
+namespace RosMessageGeneration
+{
+    public class MessageGenBrowser : EditorWindow
+    {
+        // we actually only want the hashset, but Unity won't serialize that - so we keep foldedOutList as a backup.
+        private HashSet<string> foldedOutHash;
```
[the remark this answers is not in the document]
Comment (LaurieCheers-unity) [reply Editor/MessageGeneration/MessageGenBrowser.cs:12]: Good idea.

File: Editor/MessageGeneration/MessageGenBrowser.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (mrpropellers) [reply Editor/MessageGeneration/MessageGenBrowser.cs]: Then perhaps a NotImplementedError would be appropriate?

[the remark this answers is not in the document]
Comment (LaurieCheers-unity) [reply Editor/MessageGeneration/MessageGenBrowser.cs]: Ok, I added a "hasBuildButton" property to clarify the logic, see what you think.

File: Runtime/MessageGeneration/MessageAutoGen.cs (hunk 4 above)
Comment (mrpropellers) [reply Runtime/MessageGeneration/MessageAutoGen.cs:18]: Is that logic on the way in this review or coming in a follow-up task?

Comment (LaurieCheers-unity) [reply Runtime/MessageGeneration/MessageAutoGen.cs:18]: Working on it now, but I feel like it could go in a separate PR

**Pull request merged**